File: python/dirlist.py
#!/usr/bin/env python3
"""
dirlist.py — Recursively list directories under a given path.

Demonstrates three approaches to stdout vs stderr in Python:

  DEFAULT          print() for data, print(..., file=sys.stderr) for diagnostics
  --use-logging    logging module for diagnostics (the preferred production approach)
  --wrong-output   everything via plain print() — breaks pipe consumers

Usage:
    python dirlist.py [--use-logging] [--wrong-output] <path>

Examples:
    python dirlist.py /tmp
    python dirlist.py --use-logging /tmp
    python dirlist.py /tmp | wc -l
    python dirlist.py --wrong-output /tmp | wc -l      ← broken count
    python dirlist.py --use-logging --wrong-output /tmp | wc -l  ← also broken
"""

import argparse
import logging
import os
import sys
from pathlib import Path


# ── logging setup ─────────────────────────────────────────────────────────────
#
# logging.basicConfig() defaults to stderr (StreamHandler(sys.stderr)).
# That's the correct behaviour — log output bypasses pipes automatically.
#
# The format mirrors the [INFO] / [WARN] style from the bash version so the
# two scripts are easy to compare side-by-side.

# PROG: the script's own name, included in every diagnostic line.
# In a pipeline, diagnostics from multiple programs share the same terminal —
# the name prefix is what lets you tell them apart.
PROG = Path(sys.argv[0]).name

# Exit codes — GNU convention
# 0 = success | 1 = runtime error | 2 = usage/argument error
# Note: argparse already exits with 2 on bad arguments — no extra work needed.
EXIT_OK    = 0
EXIT_ERR   = 1
EXIT_USAGE = 2

logging.basicConfig(
    level=logging.DEBUG,
    format=f"[%(levelname)s]  {PROG}: %(message)s",
    stream=sys.stderr,          # explicit, but this is already the default
)
log = logging.getLogger(__name__)


# ── emit helpers ──────────────────────────────────────────────────────────────

def emit(path: str) -> None:
    """Write a directory path to stdout — the data stream."""
    print(path)                 # print() defaults to sys.stdout


def emit_wrong(path: str) -> None:
    """Write a directory path to stdout — same as emit(), shown for symmetry."""
    print(path)


# ── diagnostic helpers (print-based) ──────────────────────────────────────────

def info_print(msg: str) -> None:
    """Correct: diagnostic → stderr via file=sys.stderr."""
    print(f"[INFO]  {PROG}: {msg}", file=sys.stderr)


def warn_print(msg: str) -> None:
    """Correct: warning → stderr via file=sys.stderr."""
    print(f"[WARN]  {PROG}: {msg}", file=sys.stderr)


def info_wrong(msg: str) -> None:
    """Wrong: diagnostic → stdout (default). Pollutes pipe consumers."""
    print(f"[INFO]  {PROG}: {msg}")     # missing file=sys.stderr  ← the mistake


def warn_wrong(msg: str) -> None:
    """Wrong: warning → stdout. Same problem (wrong stream; still includes PROG for identity)."""
    print(f"[WARN]  {PROG}: {msg}")     # missing file=sys.stderr  ← the mistake


# ── core logic ────────────────────────────────────────────────────────────────

def scan(root: Path, wrong_output: bool, use_logging: bool) -> None:
    """Walk root recursively and emit every subdirectory."""

    # Pick the right diagnostic functions based on mode
    if use_logging:
        # logging module: .info() and .warning() both go to stderr by default.
        # You never have to think about file descriptors — the handler owns that.
        info  = log.info
        warn  = log.warning
    elif wrong_output:
        # Wrong print mode: diagnostics go to stdout alongside the data.
        info  = info_wrong
        warn  = warn_wrong
    else:
        # Correct print mode: diagnostics explicitly routed to stderr.
        info  = info_print
        warn  = warn_print

    suffix = " (WRONG MODE)" if wrong_output else ""
    info(f"Starting directory scan{suffix}: {root}")

    dir_count   = 0
    skip_count  = 0

    for dirpath, dirnames, _ in os.walk(root):
        # Sort for deterministic output (os.walk order is filesystem-dependent)
        dirnames.sort()

        for dirname in dirnames:
            full = os.path.join(dirpath, dirname)

            # Simulate a permission-denied scenario, same as the bash version
            if dirname == "private":
                warn(f"Permission denied (simulated): {full}")
                skip_count += 1
                dirnames.remove(dirname)   # don't recurse into it
                continue

            emit(full)
            dir_count += 1

    info(f"Done. Found {dir_count} directories, skipped {skip_count}.")


# ── argument parsing ──────────────────────────────────────────────────────────

def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Recursively list directories. Demonstrates stdout vs stderr.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )
    parser.add_argument("path", help="Root directory to scan")
    parser.add_argument(
        "--wrong-output",
        action="store_true",
        help="Send all output (data + diagnostics) to stdout — breaks pipes",
    )
    parser.add_argument(
        "--use-logging",
        action="store_true",
        help="Use the logging module for diagnostics instead of print()",
    )
    return parser.parse_args()


# ── entry point ───────────────────────────────────────────────────────────────

def main() -> None:
    args = parse_args()

    root = Path(args.path)
    if not root.is_dir():
        # Runtime errors always go to stderr — even in wrong mode.
        # A crashed program's error message should never pollute a pipe.
        print(f"[ERROR] {PROG}: Not a directory: {root}", file=sys.stderr)
        sys.exit(EXIT_ERR)

    scan(root, wrong_output=args.wrong_output, use_logging=args.use_logging)


if __name__ == "__main__":
    main()

File: nodejs/dirlist.js
#!/usr/bin/env node
/**
 * dirlist.js — Recursively list directories under a given path.
 *
 * Demonstrates three approaches to stdout vs stderr in Node.js:
 *
 *   DEFAULT          console.log() for data, console.error() for diagnostics
 *   --use-logging    winston logger for diagnostics (the preferred production approach)
 *   --wrong-output   everything via console.log() — breaks pipe consumers
 *
 * Node.js-specific insight:
 *   Unlike Python's print(), console.error() already routes to stderr by default.
 *   The mistake in Node.js is using console.log() for EVERYTHING (data AND diagnostics).
 *   process.stdout.write() and process.stderr.write() are the lower-level primitives
 *   that console.log/error are built on top of.
 *
 * Usage:
 *   node dirlist.js [--use-logging] [--wrong-output] <path>
 *
 * Examples:
 *   node dirlist.js /tmp
 *   node dirlist.js --use-logging /tmp
 *   node dirlist.js /tmp | wc -l
 *   node dirlist.js --wrong-output /tmp | wc -l      ← broken count
 */

const fs   = require('fs');
const path = require('path');

// ── logging setup (winston) ───────────────────────────────────────────────────
//
// winston transports define WHERE log output goes.
// By default we point it at process.stderr — correct behaviour, bypasses pipes.
// Swap the transport (e.g. to a file) without touching any log call site.

const winston = require('winston');

// PROG: the script's own name, included in every diagnostic line.
// In a pipeline, diagnostics from multiple programs share the same terminal —
// the name prefix is what lets you tell them apart.
const PROG = path.basename(process.argv[1]);

// Exit codes — GNU convention
// 0 = success | 1 = runtime error | 2 = usage/argument error
const EXIT_OK    = 0;
const EXIT_ERR   = 1;
const EXIT_USAGE = 2;

const log = winston.createLogger({
  level: 'debug',
  format: winston.format.printf(({ level, message }) =>
    `[${level.toUpperCase()}]  ${PROG}: ${message}`
  ),
  transports: [
    new winston.transports.Stream({ stream: process.stderr }),  // correct: stderr
  ],
});


// ── emit helpers ──────────────────────────────────────────────────────────────

// Correct: data goes to stdout via console.log() (which wraps process.stdout.write)
const emit      = (p) => console.log(p);

// Wrong: data AND diagnostics all go through console.log() → stdout
const emitWrong = (p) => console.log(p);


// ── diagnostic helpers (console-based) ───────────────────────────────────────

// Correct: console.error() writes to stderr by default — no extra work needed.
// This is the key Node.js advantage over Python's print().
const info  = (msg) => console.error(`[INFO]  ${PROG}: ${msg}`);
const warn  = (msg) => console.error(`[WARN]  ${PROG}: ${msg}`);

// Wrong: console.log() writes to stdout — pollutes pipe consumers.
const infoWrong = (msg) => console.log(`[INFO]  ${PROG}: ${msg}`);
const warnWrong = (msg) => console.log(`[WARN]  ${PROG}: ${msg}`);


// ── recursive directory walk ──────────────────────────────────────────────────

function* walkDirs(root) {
  // fs.readdirSync with { withFileTypes: true } avoids a second stat() call per entry
  let entries;
  try {
    entries = fs.readdirSync(root, { withFileTypes: true });
  } catch {
    return;  // silently skip unreadable directories
  }

  // Sort for deterministic output
  entries.sort((a, b) => a.name.localeCompare(b.name));

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const full = path.join(root, entry.name);
    yield full;
    yield* walkDirs(full);
  }
}


// ── core logic ────────────────────────────────────────────────────────────────

function scan(root, { wrongOutput, useLogging }) {
  // Pick the right diagnostic functions based on mode
  let logInfo, logWarn;

  if (useLogging) {
    // winston: .info() and .warn() both go to stderr via the configured transport.
    // You never reference a file descriptor — the transport layer owns that.
    logInfo = (msg) => log.info(msg);
    logWarn = (msg) => log.warn(msg);
  } else if (wrongOutput) {
    // Wrong console mode: diagnostics go to stdout alongside data.
    logInfo = infoWrong;
    logWarn = warnWrong;
  } else {
    // Correct console mode: console.error() routes to stderr automatically.
    logInfo = info;
    logWarn = warn;
  }

  const suffix = wrongOutput ? ' (WRONG MODE)' : '';
  logInfo(`Starting directory scan${suffix}: ${root}`);

  let dirCount  = 0;
  let skipCount = 0;

  for (const dir of walkDirs(root)) {
    const basename = path.basename(dir);

    // Simulate a permission-denied scenario, same as the bash and Python versions
    if (basename === 'private') {
      logWarn(`Permission denied (simulated): ${dir}`);
      skipCount++;
      continue;
    }

    emit(dir);
    dirCount++;
  }

  logInfo(`Done. Found ${dirCount} directories, skipped ${skipCount}.`);
}


// ── argument parsing ──────────────────────────────────────────────────────────

const args       = process.argv.slice(2);
const wrongOutput = args.includes('--wrong-output');
const useLogging  = args.includes('--use-logging');
const searchPath  = args.find((a) => !a.startsWith('--'));

if (!searchPath) {
  // Usage error: wrong/missing arguments → stderr + exit 2
  process.stderr.write(`[ERROR] ${PROG}: Usage: node dirlist.js [--use-logging] [--wrong-output] <path>\n`);
  process.exit(EXIT_USAGE);
}

if (!fs.existsSync(searchPath) || !fs.statSync(searchPath).isDirectory()) {
  // Runtime error: valid invocation but bad input → stderr + exit 1
  process.stderr.write(`[ERROR] ${PROG}: Not a directory: ${searchPath}\n`);
  process.exit(EXIT_ERR);
}

scan(searchPath, { wrongOutput, useLogging });
